text/0000-macros-1.1.md

```diff
@@ -0,0 +1,443 @@
+and derive modes will shadow one another, with the last-mentioned
+`#![rustc_macro_crate]` attribute taking precedence.
+
+### Initial implementation details
+
+This section lays out what the initial implementation details of macros 1.1
+will look like, but none of this will be specified as a stable interface to the
+compiler. These exact details are subject to change over time as the
+requirements of the compiler change, and even amongst platforms these details
+may be subtly different.
+
+The compiler will essentially consider `rustc-macro` crates as `--crate-type
+dylib -C prefer-dyanmic`. That is, compiled the same way they are today. This
+namely means that these macros  will dynamically link to the same standard
+library as the compiler itself, therefore sharing resources like a global
+allocator, etc.
+
+The `libmacro` crate will compiled as an rlib and a static copy of it will be
+included in each macro. This crate will provide a symbol known by the compiler
+that can be dynamically loaded. The compiler will `dlopen` a macro crate in the
+same way it does today, find this symbol in `libmacro`, and call it.
+
+The `rustc_macro_define` and `rustc_macro_derive` attributes will be encoded
+into the crate's metadata, and the compiler will discover all these functions,
+load their function pointers, and pass them to the `libmacro` entry point as
+well. This provides the opportunity to register all the various expansion
+mechanisms with the compiler.
+
+The actual underlying representation of `TokenStream` will be basically the same
+as it is in the compiler today. (the details on this are a little light)
+
+## Pieces to stabilize
+
+Eventually this RFC is intended to be considered for stabilization (after it's
+implemented and proven out on nightly, of course). The summary of pieces that
+would become stable are:
+
+* The `macro` crate, and a small set of APIs within (skeleton above)
+* The `rustc-macro` crate type
+* The `#[rustc_macro_derive]` attribute
+* The `#[rustc_macro_define]` attribute (optional)
+* The signature of the `#![rustc_macro_{define,derive}]` functions
+* The `#![rustc_macro_crate]` attribute
+* Semantically being able to load macro crates compiled as `rustc-macro` into
+  the compiler, requiring that the crate was compiled by the exact compiler.
+* The semantic behavior of macros today, including:
+  * Derive modes and macros are not imported, they're just dumped into the crate
+    root.
+  * Lack of hygiene for both
+  * Shadowing behavior if you load multiple macro crates
+
+### Macros 1.1 in practice
+
+Alright, that's a lot to take in! Let's take a look at what this is all going to
+look like in practice, focusing on a case study of `#[derive(Serialize)]` for
+serde.
+
+First off, serde will provide a crate, let's call it `serde_macros`. The
+`Cargo.toml` will look like:
+
+```toml
+[package]
+name = "serde-macros"
+# ...
+
+[lib]
+rustc-macro = true
```

Comment:
Does Cargo not expose crate-type itself? Was a Cargo change necessary to make cdylibs work, for example?